### ISSUE TITLE
OverlayNG does not accept GeometryCollections with heterogeneous dimensions

### ISSUE THREAD

**Issue by jagill** [locationtech/jts#667]
It would be great if OverlayNG worked for for all valid/simple inputs, including GeometryCollections with heterogeneous dimensions.

**Comment by dr-jts** [comment]
Agreed. I have been thinking about how to define and implement this.

A suitable definition seems to be to use the distributive and associative properties of the intersection, union and difference operators to define the semantics of an expression with mixed geometries.  For implementation will probably implement the equivalent expression involving homogeneous components.   This won't be as fast as possible, but it will be understandable and faster to implement.
